airbyte-integrations/connectors/source-marketo/source_marketo/source.py

```diff
@@ -0,0 +1,530 @@
+    def stream_slices(self, sync_mode, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        """
+        Override default stream_slices CDK method to provide date_slices as page chunks for data fetch.
+        Returns list of dict, example: [{
+            "startDate": "2020-01-01T0:0:0Z",
+            "endDate": "2021-01-02T0:0:0Z"
+            },
+            {
+            "startDate": "2020-01-03T0:0:0Z",
+            "endDate": "2021-01-04T0:0:0Z"
+            },
+            ...]
+        """
+
+        start_date = pendulum.parse(self.start_date)
+        end_date = pendulum.now()
+
+        # Determine stream_state, if no stream_state we use start_date
+        if stream_state:
+            start_date = pendulum.parse(stream_state.get(self.cursor_field))
+
+        # use the lowest date between start_date and self.end_date, otherwise API fails if start_date is in future
+        start_date = min(start_date, end_date)
+        date_slices = []
+
+        while start_date <= end_date:
+            end_date_slice = start_date.add(days=self.window_in_days)
+            date_slice = {"startAt": to_datetime_str(start_date), "endAt": to_datetime_str(end_date_slice)}
+
+            date_slices.append(date_slice)
+            start_date = end_date_slice
+
+        return date_slices
+
+
+class MarketoExportBase(IncrementalMarketoStream):
+    """
+    Base class for all the streams which support bulk extract.
+    """
+
+    job_timeout = 60 * 180
+    poll_interval = 60
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    @property
+    def stream_fields(self):
+        return {}
+
+    @property
+    def stream_filter(self):
+        return {}
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return f"/bulk/v1/{self.stream_name}/export/{stream_slice['id']}/file.json"
+
+    def stream_slices(self, sync_mode, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        date_slices = super().stream_slices(sync_mode, stream_state, **kwargs)
+
+        for date_slice in date_slices:
+            param = {"fields": [], "filter": {"createdAt": date_slice}}
+            param["fields"].extend(self.stream_fields)
+            param["filter"].update(self.stream_filter)
+
+            export = next(MarketoExportCreate(self.config, stream_name=self.stream_name, param=param).read_records(sync_mode=None), {})
+
+            date_slice["id"] = export["exportId"]
+        return date_slices
+
+    def sleep_till_export_completed(self, stream_slice: Mapping[str, Any]) -> bool:
+        timeout_time = pendulum.now().add(seconds=self.job_timeout)
+        while pendulum.now() < timeout_time:
+            status = next(
+                MarketoExportStatus(self.config, stream_name=self.stream_name, export_id=stream_slice["id"]).read_records(sync_mode=None)
+            )
```

Comment:
Please hide this by some classmethod of the class MarketoExportStatus

Reply:
Updated.